### PR TITLE
Always have storage to read from summary blobs

### PR DESCRIPTION
**Pull request by jatgarg**
Container runtime bluff other layers below it that it always have storage. With this change we will have a storage which could be only blobCacheStorageService and only the internal storage would change underneath on connection to actual storage. This would also help with detached container mode where we rehydrate with snapshot with blobs and layers could ask from it with the blob ids.

### CODE REVIEW

**Comment by vladsud** [comment]
Any chance you can add some description - what's the goal of this change / what it tries to achieve?

**Comment by jatgarg** [comment]
> Any chance you can add some description - what's the goal of this change / what it tries to achieve?

Yes was in process. It was not ready for review, it just adds people automatically :)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -666,6 +666,7 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         // This code is plain wrong. It lies that it never returns undefined!!!
         // All callers should be fixed, as this API is called in detached state of container when we have
         // no storage and it's passed down the stack without right typing.
+        // back-compat 0.38 contextStorageAlwaysExists
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:669]: To be honest, I do not know what contextStorageAlwaysExists refers to

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -1276,7 +1280,7 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         this.propagateConnectionState();
     }
 
-    private async getDocumentStorageService(): Promise<IDocumentStorageService> {
+    private async connectToStorage(): Promise<IDocumentStorageService> {
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1283]: I'd just remove / inline this method - there are only two callers I believe and code would become more readable

File: packages/loader/container-loader/src/container.ts [hunk 3]
```diff
@@ -624,6 +628,8 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         });
 
         this._deltaManager = this.createDeltaManager();
+        this.internalStorageService = new NullBlobStorageService();
+        this._storageService = new BlobCacheStorageService(this.internalStorageService, this.blobsMap);
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:632]: The change itself (in isolation) looks fine, but I do not have full picture on its impact.
First of all, I believe that now this._storageService is never undefined, and thus this.storageService getter should be changed accordingly, and signature on context does not need to include undefined?

But the bigger issue is likely expectations at ContainerRuntime layer where I believe we will start getting this new storage and it gets propagated down the stack (to DDSs, I think).  I do not fully understand mechanism how we switch (and impact and propagation) today from undefined to real storage, and in particular - what layers might be caching original storage and may not see newly set storage on attachment. Maybe it's all works just fine, I just do not have good mental model of that process at the moment.

**Comment by jatgarg** [comment]
Closing this PR. Not following this route. Will create a new PR.

**Pull request closed without merging**